### PR TITLE
Test rewrite

### PR DESCRIPTION
**Pull request by jarpy**
- Use `version.txt`
- Make `Dockerfile` and `docker-compose.yml` from Jinja2 templates
- Reduce long chains of ENV var passing in the build/test process
- Run tests on the Docker host with virtualenv, not in the "tester" container; allows the tests to directly manipulate the containers
- Use Testinfra for its nice assertions about files, processes, users, etc.
- Allow invocations of docker and testinfra with having to go through the Makefile (very handy for running individual tests with `./bin/testinfra`)
- Use non-master node configuration and default replica shards to eliminate split-brain and "yellow" cluster problems; removes the need for `$ES_NODE_COUNT`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,59 +1,70 @@
...
-run-es-single:
-	ES_NODE_COUNT=1 docker-compose -f docker-compose.yml -f docker-compose.hostports.yml up --build elasticsearch1
+pristine: clean
+	docker rmi -f $(VERSIONED_IMAGE)
 
-run-es-cluster:
-	ES_NODE_COUNT=2 docker-compose -f docker-compose.yml -f docker-compose.hostports.yml up --build elasticsearch1 elasticsearch2
+run: run-single
 
-single-node-test: export ES_NODE_COUNT=1
```
Comment (dliappis) [inline Makefile]: One thing that I would still like to have is two test runs one for the cluster and one with ES in single mode (stand alone).
The cluster should sufficiently cover most test cases, but you'd be surprised how useful single node tests have been in uncovering issues so far.

**Comment by jarpy** [comment]
Whoops. Sorry I broke some things. It's _extremely_ kind of you to fix them!

File: Makefile [hunk 2]
```diff
@@ -1,59 +1,70 @@
...
+test: lint build docker-compose.yml
+	./bin/testinfra tests
+
+lint: venv
+	flake8 tests
 
-# Clean up left over containers and volumes from earlier failed runs
 clean:
-	docker-compose down -v && docker-compose rm -f -v
+	docker-compose down -v
```
Comment (jarpy) [inline Makefile]: We could also just make `clean:` depend on `docker-compose.yml:`, which it will then delete. Sneaky!

File: Makefile (hunk 1 above)
Comment (jarpy) [inline Makefile]: Cool. Thanks for finding such a nice way to do this. My goal was to eliminate magic environment variables, not to make the tests less comprehensive!

File: Makefile [hunk 3]
```diff
@@ -1,59 +1,70 @@
...
-	ES_NODE_COUNT=2 docker-compose -f docker-compose.yml -f docker-compose.hostports.yml up --build elasticsearch1 elasticsearch2
+run: run-single
 
-single-node-test: export ES_NODE_COUNT=1
-single-node-test: clean
-	docker-compose up -d --build elasticsearch1
-	docker-compose build --pull tester
-	docker-compose run tester
-	docker-compose down -v
+run-single: build
```
Comment (jarpy) [inline Makefile]: Yep. Didn't test those targets. 😳 Thanks.

**Pull request merged**